lib/active_merchant/billing/gateways/pin.rb

```diff
@@ -178,10 +184,9 @@ def add_metadata(post, options)
       def add_3ds(post, options)
         if options[:three_d_secure]
           post[:three_d_secure] = {}
-          post[:three_d_secure][:version] = options[:three_d_secure][:version] if options[:three_d_secure][:version]
-          post[:three_d_secure][:eci] = options[:three_d_secure][:eci] if options[:three_d_secure][:eci]
-          post[:three_d_secure][:cavv] = options[:three_d_secure][:cavv] if options[:three_d_secure][:cavv]
-          post[:three_d_secure][:transaction_id] = options[:three_d_secure][:ds_transaction_id] || options[:three_d_secure][:xid]
+          post[:three_d_secure][:enabled] = options[:three_d_secure][:enabled] if options[:three_d_secure][:enabled]
+          post[:three_d_secure][:fallback_ok] = options[:three_d_secure][:fallback_ok] if options[:three_d_secure][:fallback_ok]
+          post[:three_d_secure][:callback_url] = options[:three_d_secure][:callback_url] if options[:three_d_secure][:callback_url]
```

Comment:
@hudakh this will still need to support the old style of 3DS as implemented in https://github.com/activemerchant/active_merchant/pull/3848, I'd suggest doing something like this:

```ruby
def add_3ds(post, options)
  if options[:three_d_secure]
    post[:three_d_secure] = {}

    if options[:three_d_secure][:enabled]
      post[:three_d_secure][:enabled] = true
      post[:three_d_secure][:fallback_ok] = options[:three_d_secure][:fallback_ok] if options[:three_d_secure][:fallback_ok]
      post[:three_d_secure][:callback_url] = options[:three_d_secure][:callback_url] if options[:three_d_secure][:callback_url]
    else
      post[:three_d_secure][:version] = options[:three_d_secure][:version] if options[:three_d_secure][:version]
      post[:three_d_secure][:eci] = options[:three_d_secure][:eci] if options[:three_d_secure][:eci]
      post[:three_d_secure][:cavv] = options[:three_d_secure][:cavv] if options[:three_d_secure][:cavv]
      post[:three_d_secure][:transaction_id] = options[:three_d_secure][:ds_transaction_id] || options[:three_d_secure][:xid]
    end
  end
end
```

Reply:
Thanks @AMHOL. I've added a commit with your suggestion.